RecoLocalCalo/EcalRecProducers/plugins/EcalDetailedTimeRecHitProducer.cc

```diff
@@ -0,0 +1,240 @@
+   
+  correctForVertexZPosition_ = ps.getParameter<bool>("correctForVertexZPosition");
+  useMCTruthVertex_ = ps.getParameter<bool>("useMCTruthVertex");
+  recoVertex_       = consumes<reco::VertexCollection>( ps.getParameter<edm::InputTag>("recoVertex") );
+  simVertex_       = consumes<edm::SimVertexContainer>( ps.getParameter<edm::InputTag>("simVertex") );
+  
+  ebTimeLayer_ = ps.getParameter<int>("EBTimeLayer");
+  eeTimeLayer_ = ps.getParameter<int>("EETimeLayer");
+
+  produces< EBRecHitCollection >(EBDetailedTimeRecHitCollection_);
+  produces< EERecHitCollection >(EEDetailedTimeRecHitCollection_);
+}
+
+EcalDetailedTimeRecHitProducer::~EcalDetailedTimeRecHitProducer() {
+}
+
+
+void EcalDetailedTimeRecHitProducer::produce(edm::Event& evt, const edm::EventSetup& es)
+{
+        using namespace edm;
+        using namespace reco;
+
+	edm::ESHandle<CaloGeometry>               hGeometry   ;
+	es.get<CaloGeometryRecord>().get( hGeometry ) ;
+	
+	m_geometry = &*hGeometry;
+
+        Handle< EBRecHitCollection > pEBRecHits;
+        Handle< EERecHitCollection > pEERecHits;
+
+        const EBRecHitCollection*  EBRecHits = 0;
+        const EERecHitCollection*  EERecHits = 0; 
+
+	evt.getByToken( EBRecHitCollection_, pEBRecHits);
+	if ( pEBRecHits.isValid() ) {
+	  EBRecHits = pEBRecHits.product(); // get a ptr to the product
+#ifdef DEBUG
+	  LogDebug("EcalRecHitDebug") << "total # EB rechits to be re-calibrated: " << EBRecHits->size();
+#endif
+	}
+        
+	evt.getByToken( EERecHitCollection_, pEERecHits);
+	if ( pEERecHits.isValid() ) {
+	  EERecHits = pEERecHits.product(); // get a ptr to the product
+#ifdef DEBUG
+	  LogDebug("EcalRecHitDebug") << "total # EE uncalibrated rechits to be re-calibrated: " << EERecHits->size();
+#endif
+	} 
+        
+        Handle< EcalTimeDigiCollection > pEBTimeDigis;
+        Handle< EcalTimeDigiCollection > pEETimeDigis;
+
+        const EcalTimeDigiCollection* ebTimeDigis =0;
+        const EcalTimeDigiCollection* eeTimeDigis =0;
+
+	evt.getByToken( ebTimeDigiCollection_, pEBTimeDigis);
+	//evt.getByToken( digiProducer_, pEBTimeDigis);
+	if ( pEBTimeDigis.isValid() ) {
+	  ebTimeDigis = pEBTimeDigis.product(); // get a ptr to the produc
+	  edm::LogInfo("EcalDetailedTimeRecHitInfo") << "total # ebTimeDigis: " << ebTimeDigis->size() ;
+	} 
+
+	evt.getByToken( eeTimeDigiCollection_, pEETimeDigis);
+	//evt.getByToken( digiProducer_, pEETimeDigis);
+	if ( pEETimeDigis.isValid() ) {
+	  eeTimeDigis = pEETimeDigis.product(); // get a ptr to the product
+	  edm::LogInfo("EcalDetailedTimeRecHitInfo") << "total # eeTimeDigis: " << eeTimeDigis->size() ;
+	}         
+        // collection of rechits to put in the event
+        std::unique_ptr< EBRecHitCollection > EBDetailedTimeRecHits( new EBRecHitCollection );
+        std::unique_ptr< EERecHitCollection > EEDetailedTimeRecHits( new EERecHitCollection );
+
+	GlobalPoint* vertex=0;
```

Comment:
It would probably be safer to use unique_ptr here in case of thrown exceptions or future changes to this code that might add `return` statements.